framework/src/vectorpostprocessors/WorkBalance.C

```diff
@@ -0,0 +1,225 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "WorkBalance.h"
+
+// MOOSE includes
+#include "MooseVariableField.h"
+#include "ThreadedElementLoopBase.h"
+#include "ThreadedNodeLoop.h"
+
+#include "libmesh/quadrature.h"
+
+registerMooseObject("MooseApp", WorkBalance);
+
+template <>
+InputParameters
+validParams<WorkBalance>()
+{
+  InputParameters params = validParams<GeneralVectorPostprocessor>();
+  params.addClassDescription("Computes several metrics for workload balance per processor");
+  return params;
+}
+
+WorkBalance::WorkBalance(const InputParameters & parameters)
+  : GeneralVectorPostprocessor(parameters),
+    _local_num_elems(0),
+    _local_num_nodes(0),
+    _local_num_dofs(0),
+    _local_num_partition_sides(0),
+    _local_partition_surface_area(0),
+    _pid(declareVector("pid")),
+    _num_elems(declareVector("num_elems")),
+    _num_nodes(declareVector("num_nodes")),
+    _num_dofs(declareVector("num_dofs")),
+    _num_partition_sides(declareVector("num_partition_sides")),
+    _partition_surface_area(declareVector("partition_surface_area"))
+{
+}
+
+void
+WorkBalance::initialize()
+{
+  _local_num_elems = 0;
+  _local_num_nodes = 0;
+  _local_num_dofs = 0;
+  _local_num_partition_sides = 0;
+  _local_partition_surface_area = 0;
+}
+
+namespace
+{
+
+// Helper Threaded Loop for Elements
+class WBElementLoop : public ThreadedElementLoopBase<ConstElemRange>
+{
+public:
+  WBElementLoop(MooseMesh & mesh)
+    : ThreadedElementLoopBase(mesh),
+      _local_num_elems(0),
+      _local_num_dofs(0),
+      _local_num_partition_sides(0),
+      _local_partition_surface_area(0),
+      _this_pid(_mesh.processor_id()) // Get this once because it is expensive
+  {
+  }
+
+  WBElementLoop(WBElementLoop & x, Threads::split split)
+    : ThreadedElementLoopBase(x, split),
+      _local_num_elems(0),
+      _local_num_dofs(0),
+      _local_num_partition_sides(0),
+      _local_partition_surface_area(0),
+      _this_pid(x._this_pid)
+  {
+  }
+
+  virtual ~WBElementLoop() {}
+
+  virtual void pre() override
+  {
+    _local_num_elems = 0;
+    _local_num_dofs = 0;
+    _local_num_partition_sides = 0;
+    _local_partition_surface_area = 0;
+  }
+
+  virtual void onElement(const Elem * elem) override
+  {
+    _local_num_elems++;
+
+    /*
+    // Find out how many dofs there are on this element
+    auto n_sys = elem->n_systems();
+    for (decltype(n_sys) sys = 0; sys < n_sys; sys++)
+    {
+    */
+
+    // For MOOSE, system 0 is the nonnlinear system - that's what we care about here
+    // I've left the other code commented out here because I might change my mind in a little while
+    // and add back the ability to set the system or compute over all
```

Comment:
Perhaps an option for either or both is in order: https://github.com/idaholab/moose/blob/215da4b843f8a5240ac5f936159bc22f53ee5f68/framework/src/postprocessors/NumDOFs.C#L20

Reply:
Yeah, I'm just commenting that we previously ran into this before and I read your comment here showing you that we eventually added the option. You don't have to change anything, but you can see that we may want to add options. Again, I'm not saying that you should change it, just getting you to think and be consistent with other pars of the code.

Reply:
Well - the other problem here is that it might be some other system other than NL or Aux.  For instance: I have a few systems in my ray tracing stuff.

I know that I can get the enum to accept more values than what are given... but how exactly do I get an _id_ for those?  Do you have to give the real name of the system (like `rt0`) or do you provide the system number or something?

This is what I wasn't quite sure of so I didn't know how I wanted to proceed.

I'll at least add in `NL`, `AUX` and `ALL` for now.

Reply:
Also: thanks for pointing to `NumDOFs`... I'll make this consistent.

Reply:
I'm not sure how we get the extra ids. @aeslaughter - has rewritten that system and things have changed quite a bit. NL and Aux will be sufficient for almost anyone